src/H5Iint.c

```diff
@@ -8811,505 +8866,162 @@ H5I__discard_mt_id_info(H5I_mt_id_info_t * id_info_ptr)
+                    /* set up new_fl_shead */
+                    assert(fl_snext.ptr);
+                    new_fl_shead.ptr = fl_snext.ptr;
+                    new_fl_shead.sn  = fl_shead.sn + 1;
 
-                }
-            } else {
+                    if ( ! atomic_compare_exchange_strong(&(H5I_mt_g.id_info_fl_shead), 
+                                                          &fl_shead, new_fl_shead) ) {
 
-                // assert(LFHT_FL_NODE_ON_FL == atomic_load(&(fl_next->tag)));
+                        /* the attempt to remove the first item from the free list
+                         * failed.  Update stats and try again.
+                         */
+                        atomic_fetch_add(&(H5I_mt_g.num_id_info_fl_head_update_cols), 1ULL);
 
-                /* attempt to set lfht_ptr->fl_stail to fl_next.  It doesn't
-                 * matter whether we succeed or fail, as if we fail, it
-                 * just means that some other thread beat us to it.
-                 *
-                 * that said, it doesn't hurt to collect stats
-                 */
-                new_fl_stail.ptr = fl_snext.ptr;
-                new_fl_stail.sn  = fl_stail.sn + 1;
-                if ( ! atomic_compare_exchange_strong(&(H5I_mt_g.id_info_fl_stail), &fl_stail, new_fl_stail) ) {
+                    } else {
 
-                    atomic_fetch_add(&(H5I_mt_g.num_id_info_fl_tail_update_cols), 1);
-                }
-            }
-        }
-    }
+                        H5I_mt_id_info_sptr_t null_snext = {NULL, 0ULL};
 
-    /* test to see if both H5I_mt_g.id_info_fl_len and H5I_mt_g.num_id_info_fl_entries_reallocable 
-     * are greater than H5I_mt_g.max_desired_id_info_fl_len.  Check both, as while we want to 
-     * keep the free list length around H5I_mt_g.max_desired_id_info_fl_len, we also want to 
-     * have an ample supply of reallocable free list entries on hand.
-     */
-    assert(atomic_load(&(H5I_mt_g.num_id_info_fl_entries_reallocable)) <= 
-           atomic_load(&(H5I_mt_g.id_info_fl_len)));
+                        /* first has been removed from the free list.  Set id_info_ptr to fl_shead.ptr,
+                         * discard *id_info_ptr, update stats, and exit the loop by setting done to true.
+                         */
+                        id_info_ptr = fl_shead.ptr;
 
-    if ( ( atomic_load(&(H5I_mt_g.id_info_fl_len)) > atomic_load(&(H5I_mt_g.max_desired_id_info_fl_len)) ) &&
-         ( atomic_load(&(H5I_mt_g.num_id_info_fl_entries_reallocable)) > 
-           atomic_load(&(H5I_mt_g.max_desired_id_info_fl_len)) ) ) {
+                        assert(H5I__ID_INFO == id_info_ptr->tag);
+                        assert(atomic_load(&(id_info_ptr->on_fl)));
+                        assert(0 < atomic_load(&(id_info_ptr->serial_num)));
 
-        try_to_free_an_entry = TRUE;
+                        /* the above assert only exists in production builds.  If id_info_ptr->serial_num
```

Comment:
I'm not sure what is meant by `the above assert only exists in production builds`. Asserts exist when NDEBUG is not defined, which is only in debug builds.

Reply:
I'm also not sure exactly what that comment means, John wrote that and I'll ask him for an explanation and I will update it.

Reply:
If the Assert doesn't trigger, at least we log the issue.